rust/sbd-bench/src/c_count_scale.rs

```diff
@@ -0,0 +1,166 @@
+use super::*;
+use std::sync::Mutex;
+
+pub struct Fail(Arc<tokio::sync::Semaphore>, bool);
+
+impl Clone for Fail {
+    fn clone(&self) -> Self {
+        Self(self.0.clone(), false)
+    }
+}
+
+impl Drop for Fail {
+    fn drop(&mut self) {
+        if self.1 {
+            self.0.close();
+        }
+    }
+}
+
+impl Default for Fail {
+    fn default() -> Self {
+        Self(Arc::new(tokio::sync::Semaphore::new(0)), false)
+    }
+}
+
+impl Fail {
+    fn set_fail_on_drop(&mut self, fail_on_drop: bool) {
+        self.1 = fail_on_drop;
+    }
+
+    async fn fail(&self) {
+        let _ = self.0.acquire().await;
+    }
+}
+
+struct Stats {
+    addr: std::net::SocketAddr,
+    last_ip: u32,
+    client_count: usize,
+    messages_sent: usize,
+}
+
+impl std::fmt::Debug for Stats {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Stats")
+            .field("client_count", &self.client_count)
+            .field("messages_sent", &self.messages_sent)
+            .finish()
+    }
+}
+
+async fn create_client(this: &Mutex<Stats>) -> (SbdClient, MsgRecv) {
+    let crypto = DefaultCrypto::default();
+
+    let (ip, addr) = {
+        let mut lock = this.lock().unwrap();
+        let ip = lock.last_ip;
+        lock.last_ip += 1;
+        lock.client_count += 1;
+        (ip, lock.addr)
+    };
+
+    let ip = ip.to_be_bytes();
+    let ip = std::net::Ipv4Addr::new(ip[0], ip[1], ip[2], ip[3]);
+    let ip = format!("{ip}");
+
+    let c = SbdClientConfig {
+        allow_plain_text: true,
+        headers: vec![("test-ip".to_string(), ip)],
+        ..Default::default()
+    };
+
+    let url = format!("ws://{}", addr);
+
+    SbdClient::connect_config(&url, &crypto, c).await.unwrap()
+}
+
+pub async fn c_count_scale(max: usize) -> ! {
+    let config = Arc::new(Config {
+        bind: vec!["127.0.0.1:0".to_string(), "[::1]:0".to_string()],
+        trusted_ip_header: Some("test-ip".to_string()),
+        ..Default::default()
+    });
+
+    let server = SbdServer::new(config).await.unwrap();
+
+    let addr = *server.bind_addrs().first().unwrap();
+
+    let stats = Arc::new(Mutex::new(Stats {
+        addr,
+        last_ip: u32::from_be_bytes([1, 1, 1, 1]),
+        client_count: 0,
+        messages_sent: 0,
+    }));
+
+    let fail = Fail::default();
+
+    for _ in 0..16 {
+        let stats = stats.clone();
+        let mut fail = fail.clone();
+        tokio::task::spawn(async move {
+            loop {
+                fail.set_fail_on_drop(true);
+
+                let client_count = stats.lock().unwrap().client_count;
+                if client_count > max {
```

Comment:
So the `max` is like a 'soft' maximum where it'll keep creating clients but those connections are allowed to fail without failing the test?

What is the actual behavior like - do some of the connections fail when the server starts to get busy so it's useful to allow this, or do connections generally succeed and this is just here to get a useful test result when the server does hit its limits?

Reply:
The outer task loop that is establishing new connections exits without failing the test, but any inner tasks (with communicating connection pairs) continue running unless they error, in which case they will still fail the test.